Stack trace from a frozen run:

```
panic: test timed out after 10m0s
                                                                                                         
goroutine 161 [running]:                                                                                                                                                                                           
testing.(*M).startAlarm.func1()                                                                          
        /usr/local/google/home/dbsmith/.gvm/gos/go1.16/src/testing/testing.go:1701 +0xe5                                                                                                                           
created by time.goFunc
        /usr/local/google/home/dbsmith/.gvm/gos/go1.16/src/time/sleep.go:180 +0x45
                                                                                                         
goroutine 1 [chan receive]:                                                                                                                                                                                        
testing.(*T).Run(0xc000001c80, 0x6a3c45, 0x16, 0x6b2d40, 0x490f01)
        /usr/local/google/home/dbsmith/.gvm/gos/go1.16/src/testing/testing.go:1240 +0x2da                                                                                                                          
testing.runTests.func1(0xc000102900)
        /usr/local/google/home/dbsmith/.gvm/gos/go1.16/src/testing/testing.go:1512 +0x78
testing.tRunner(0xc000102900, 0xc000151d68)                                                              
        /usr/local/google/home/dbsmith/.gvm/gos/go1.16/src/testing/testing.go:1194 +0xef                                                                                                                           
testing.runTests(0xc0001380c0, 0x84e780, 0x14, 0x14, 0xc03dc8a4d8509dbf, 0x8bb2d5261e, 0x853820, 0x203000)
        /usr/local/google/home/dbsmith/.gvm/gos/go1.16/src/testing/testing.go:1510 +0x2fe                                                                                                                          
testing.(*M).Run(0xc0001c2000, 0x0)
        /usr/local/google/home/dbsmith/.gvm/gos/go1.16/src/testing/testing.go:1418 +0x1eb
k8s.io/client-go/util/workqueue.TestMain(0xc0001c2000)                   
        /usr/local/google/home/dbsmith/code/k4/staging/src/k8s.io/client-go/util/workqueue/main_test.go:28 +0x93       
main.main()                                                                                              
        _testmain.go:89 +0x165                                                                                                                                                                                     

goroutine 19 [chan receive]:
k8s.io/klog/v2.(*loggingT).flushDaemon(0x853980)                                                         
        /usr/local/google/home/dbsmith/.gvm/pkgsets/go1.16/global/pkg/mod/k8s.io/klog/v2@v2.9.0/klog.go:1169 +0x8b     
created by k8s.io/klog/v2.init.0                                                                         
        /usr/local/google/home/dbsmith/.gvm/pkgsets/go1.16/global/pkg/mod/k8s.io/klog/v2@v2.9.0/klog.go:420 +0xdf     

goroutine 10 [sync.Cond.Wait]:
sync.runtime_notifyListWait(0xc0000122d0, 0xc000000002)                  
        /usr/local/google/home/dbsmith/.gvm/gos/go1.16/src/runtime/sema.go:513 +0xf8                                                                                                                               
sync.(*Cond).Wait(0xc0000122c0)                                                                          
        /usr/local/google/home/dbsmith/.gvm/gos/go1.16/src/sync/cond.go:56 +0x99                                                                                                                                   
k8s.io/client-go/util/workqueue.(*Type).wait(0xc000030a20)
        /usr/local/google/home/dbsmith/code/k4/staging/src/k8s.io/client-go/util/workqueue/queue.go:224 +0x7a
k8s.io/client-go/util/workqueue.(*Type).ShutDownWithDrain(0xc000030a20)
        /usr/local/google/home/dbsmith/code/k4/staging/src/k8s.io/client-go/util/workqueue/queue.go:209 +0x50
k8s.io/client-go/util/workqueue_test.TestAddWhileProcessing(0xc000001c80)
        /usr/local/google/home/dbsmith/code/k4/staging/src/k8s.io/client-go/util/workqueue/queue_test.go:143 +0x1ac
testing.tRunner(0xc000001c80, 0x6b2d40)
        /usr/local/google/home/dbsmith/.gvm/gos/go1.16/src/testing/testing.go:1194 +0xef
created by testing.(*T).Run
        /usr/local/google/home/dbsmith/.gvm/gos/go1.16/src/testing/testing.go:1239 +0x2b3

goroutine 25 [select]:
k8s.io/client-go/util/workqueue.(*delayingType).waitingLoop(0xc00012a480)
        /usr/local/google/home/dbsmith/code/k4/staging/src/k8s.io/client-go/util/workqueue/delaying_queue.go:231 +0x3df
created by k8s.io/client-go/util/workqueue.newDelayingQueue
        /usr/local/google/home/dbsmith/code/k4/staging/src/k8s.io/client-go/util/workqueue/delaying_queue.go:68 +0x185

goroutine 31 [select]:
k8s.io/client-go/util/workqueue.(*delayingType).waitingLoop(0xc00012ae40)
        /usr/local/google/home/dbsmith/code/k4/staging/src/k8s.io/client-go/util/workqueue/delaying_queue.go:231 +0x3df
created by k8s.io/client-go/util/workqueue.newDelayingQueue
        /usr/local/google/home/dbsmith/code/k4/staging/src/k8s.io/client-go/util/workqueue/delaying_queue.go:68 +0x185

goroutine 9 [chan receive]:
k8s.io/client-go/util/workqueue.(*Type).updateUnfinishedWorkLoop(0xc0000303c0)
        /usr/local/google/home/dbsmith/code/k4/staging/src/k8s.io/client-go/util/workqueue/queue.go:256 +0xac
created by k8s.io/client-go/util/workqueue.newQueue
        /usr/local/google/home/dbsmith/code/k4/staging/src/k8s.io/client-go/util/workqueue/queue.go:63 +0x154

goroutine 36 [select]:
k8s.io/client-go/util/workqueue.(*delayingType).waitingLoop(0xc00020e4e0)
        /usr/local/google/home/dbsmith/code/k4/staging/src/k8s.io/client-go/util/workqueue/delaying_queue.go:231 +0x3df
created by k8s.io/client-go/util/workqueue.newDelayingQueue
        /usr/local/google/home/dbsmith/code/k4/staging/src/k8s.io/client-go/util/workqueue/delaying_queue.go:68 +0x185

goroutine 52 [select]:
k8s.io/client-go/util/workqueue.(*delayingType).waitingLoop(0xc00012b6e0)
        /usr/local/google/home/dbsmith/code/k4/staging/src/k8s.io/client-go/util/workqueue/delaying_queue.go:231 +0x3df
created by k8s.io/client-go/util/workqueue.newDelayingQueue
        /usr/local/google/home/dbsmith/code/k4/staging/src/k8s.io/client-go/util/workqueue/delaying_queue.go:68 +0x185

goroutine 99 [select]:
k8s.io/client-go/util/workqueue.(*delayingType).waitingLoop(0xc00020e8a0)
        /usr/local/google/home/dbsmith/code/k4/staging/src/k8s.io/client-go/util/workqueue/delaying_queue.go:231 +0x3df
created by k8s.io/client-go/util/workqueue.newDelayingQueue
        /usr/local/google/home/dbsmith/code/k4/staging/src/k8s.io/client-go/util/workqueue/delaying_queue.go:68 +0x185
FAIL    k8s.io/client-go/util/workqueue 600.032s
```

--- a/pkg/controller/garbagecollector/garbagecollector_test.go
+++ b/pkg/controller/garbagecollector/garbagecollector_test.go
@@ -2767,6 +2767,9 @@ func (t *trackingWorkqueue) Len() int {
 func (t *trackingWorkqueue) ShutDown() {
 	t.limiter.ShutDown()
 }
+func (t *trackingWorkqueue) ShutDownWithDrain() {
+	t.limiter.ShutDownWithDrain()
+}
 func (t *trackingWorkqueue) ShuttingDown() bool {
 	return t.limiter.ShuttingDown()
 }

--- a/staging/src/k8s.io/client-go/util/workqueue/queue.go
+++ b/staging/src/k8s.io/client-go/util/workqueue/queue.go
@@ -29,6 +29,7 @@ type Interface interface {
 	Get() (item interface{}, shutdown bool)
 	Done(item interface{})
 	ShutDown()
+	ShutDownWithDrain()
 	ShuttingDown() bool
 }
 
@@ -86,6 +87,7 @@ type Type struct {
 	cond *sync.Cond
 
 	shuttingDown bool
+	drain        bool
 
 	metrics queueMetrics
 
@@ -108,6 +110,10 @@ func (s set) insert(item t) {
 
 func (s set) delete(item t) {
 	delete(s, item)
+}
+
+func (s set) len() int {
+	return len(s)
 }
 
 // Add marks item as needing processing.
@@ -178,13 +184,71 @@ func (q *Type) Done(item interface{}) {
 	if q.dirty.has(item) {
 		q.queue = append(q.queue, item)
 		q.cond.Signal()
+	} else if q.processing.len() == 0 {
+		q.cond.Signal()
 	}
 }
 
-// ShutDown will cause q to ignore all new items added to it. As soon as the
-// worker goroutines have drained the existing items in the queue, they will be
-// instructed to exit.
+// ShutDown will cause q to ignore all new items added to it and
+// immediately instruct the worker goroutines to exit.
 func (q *Type) ShutDown() {
+	q.setDrain(false)
+	q.shutdown()
+}
+
+// ShutDownWithDrain will cause q to ignore all new items added to it. As soon
+// as the worker goroutines have "drained", i.e: finished processing and called
+// Done on all existing items in the queue; they will be instructed to exit and
+// ShutDownWithDrain will return. Hence: a strict requirement for using this is;
+// your workers must ensure that Done is called on all items in the queue once
+// the shut down has been initiated, if that is not the case: this will block
+// indefinitely. It is, however, safe to call ShutDown after having called
+// ShutDownWithDrain, as to force the queue shut down to terminate immediately
+// without waiting for the drainage.
+func (q *Type) ShutDownWithDrain() {
+	q.setDrain(true)
+	q.shutdown()
+	for q.isProcessing() && q.shouldDrain() {
+		q.waitForProcessing()
+	}
+}
+
+// isProcessing indicates if there are still items on the work queue being
+// processed. It's used to drain the work queue on an eventual shutdown.
+func (q *Type) isProcessing() bool {
+	q.cond.L.Lock()
+	defer q.cond.L.Unlock()
+	return q.processing.len() != 0
+}
+
+// waitForProcessing waits for the worker goroutines to finish processing items
+// and call Done on them.
+func (q *Type) waitForProcessing() {
+	q.cond.L.Lock()
+	defer q.cond.L.Unlock()
+	// Ensure that we do not wait on a queue which is already empty, as that
+	// could result in waiting for Done to be called on items in an empty queue
+	// which has already been shut down, which will result in waiting
+	// indefinitely.
+	if q.processing.len() == 0 {
+		return
+	}
+	q.cond.Wait()
+}
+
+func (q *Type) setDrain(shouldDrain bool) {
+	q.cond.L.Lock()
+	defer q.cond.L.Unlock()
+	q.drain = shouldDrain
+}
+
+func (q *Type) shouldDrain() bool {
+	q.cond.L.Lock()
+	defer q.cond.L.Unlock()
+	return q.drain
+}
+
+func (q *Type) shutdown() {
 	q.cond.L.Lock()
 	defer q.cond.L.Unlock()
 	q.shuttingDown = true

--- a/staging/src/k8s.io/client-go/util/workqueue/queue_test.go
+++ b/staging/src/k8s.io/client-go/util/workqueue/queue_test.go
@@ -25,93 +25,127 @@ import (
 )
 
 func TestBasic(t *testing.T) {
-	// If something is seriously wrong this test will never complete.
-	q := workqueue.New()
-
-	// Start producers
-	const producers = 50
-	producerWG := sync.WaitGroup{}
-	producerWG.Add(producers)
-	for i := 0; i < producers; i++ {
-		go func(i int) {
-			defer producerWG.Done()
-			for j := 0; j < 50; j++ {
-				q.Add(i)
-				time.Sleep(time.Millisecond)
-			}
-		}(i)
+	tests := []struct {
+		queue         *workqueue.Type
+		queueShutDown func(workqueue.Interface)
+	}{
+		{
+			queue:         workqueue.New(),
+			queueShutDown: workqueue.Interface.ShutDown,
+		},
+		{
+			queue:         workqueue.New(),
+			queueShutDown: workqueue.Interface.ShutDownWithDrain,
+		},
 	}
+	for _, test := range tests {
+		// If something is seriously wrong this test will never complete.
 
-	// Start consumers
-	const consumers = 10
-	consumerWG := sync.WaitGroup{}
-	consumerWG.Add(consumers)
-	for i := 0; i < consumers; i++ {
-		go func(i int) {
-			defer consumerWG.Done()
-			for {
-				item, quit := q.Get()
-				if item == "added after shutdown!" {
-					t.Errorf("Got an item added after shutdown.")
+		// Start producers
+		const producers = 50
+		producerWG := sync.WaitGroup{}
+		producerWG.Add(producers)
+		for i := 0; i < producers; i++ {
+			go func(i int) {
+				defer producerWG.Done()
+				for j := 0; j < 50; j++ {
+					test.queue.Add(i)
+					time.Sleep(time.Millisecond)
 				}
-				if quit {
-					return
+			}(i)
+		}
+
+		// Start consumers
+		const consumers = 10
+		consumerWG := sync.WaitGroup{}
+		consumerWG.Add(consumers)
+		for i := 0; i < consumers; i++ {
+			go func(i int) {
+				defer consumerWG.Done()
+				for {
+					item, quit := test.queue.Get()
+					if item == "added after shutdown!" {
+						t.Errorf("Got an item added after shutdown.")
+					}
+					if quit {
+						return
+					}
+					t.Logf("Worker %v: begin processing %v", i, item)
+					time.Sleep(3 * time.Millisecond)
+					t.Logf("Worker %v: done processing %v", i, item)
+					test.queue.Done(item)
 				}
-				t.Logf("Worker %v: begin processing %v", i, item)
-				time.Sleep(3 * time.Millisecond)
-				t.Logf("Worker %v: done processing %v", i, item)
-				q.Done(item)
-			}
-		}(i)
+			}(i)
+		}
+
+		producerWG.Wait()
+		test.queueShutDown(test.queue)
+		test.queue.Add("added after shutdown!")
+		consumerWG.Wait()
+		if test.queue.Len() != 0 {
+			t.Errorf("Expected the queue to be empty, had: %v items", test.queue.Len())
+		}
 	}
-
-	producerWG.Wait()
-	q.ShutDown()
-	q.Add("added after shutdown!")
-	consumerWG.Wait()
 }
 
 func TestAddWhileProcessing(t *testing.T) {
-	q := workqueue.New()
-
-	// Start producers
-	const producers = 50
-	producerWG := sync.WaitGroup{}
-	producerWG.Add(producers)
-	for i := 0; i < producers; i++ {
-		go func(i int) {
-			defer producerWG.Done()
-			q.Add(i)
-		}(i)
+	tests := []struct {
+		queue         *workqueue.Type
+		queueShutDown func(workqueue.Interface)
+	}{
+		{
+			queue:         workqueue.New(),
+			queueShutDown: workqueue.Interface.ShutDown,
+		},
+		{
+			queue:         workqueue.New(),
+			queueShutDown: workqueue.Interface.ShutDownWithDrain,
+		},
 	}
+	for _, test := range tests {
 
-	// Start consumers
-	const consumers = 10
-	consumerWG := sync.WaitGroup{}
-	consumerWG.Add(consumers)
-	for i := 0; i < consumers; i++ {
-		go func(i int) {
-			defer consumerWG.Done()
-			// Every worker will re-add every item up to two times.
-			// This tests the dirty-while-processing case.
-			counters := map[interface{}]int{}
-			for {
-				item, quit := q.Get()
-				if quit {
-					return
+		// Start producers
+		const producers = 50
+		producerWG := sync.WaitGroup{}
+		producerWG.Add(producers)
+		for i := 0; i < producers; i++ {
+			go func(i int) {
+				defer producerWG.Done()
+				test.queue.Add(i)
+			}(i)
+		}
+
+		// Start consumers
+		const consumers = 10
+		consumerWG := sync.WaitGroup{}
+		consumerWG.Add(consumers)
+		for i := 0; i < consumers; i++ {
+			go func(i int) {
+				defer consumerWG.Done()
+				// Every worker will re-add every item up to two times.
+				// This tests the dirty-while-processing case.
+				counters := map[interface{}]int{}
+				for {
+					item, quit := test.queue.Get()
+					if quit {
+						return
+					}
+					counters[item]++
+					if counters[item] < 2 {
+						test.queue.Add(item)
+					}
+					test.queue.Done(item)
 				}
-				counters[item]++
-				if counters[item] < 2 {
-					q.Add(item)
-				}
-				q.Done(item)
-			}
-		}(i)
+			}(i)
+		}
+
+		producerWG.Wait()
+		test.queueShutDown(test.queue)
+		consumerWG.Wait()
+		if test.queue.Len() != 0 {
+			t.Errorf("Expected the queue to be empty, had: %v items", test.queue.Len())
+		}
 	}
-
-	producerWG.Wait()
-	q.ShutDown()
-	consumerWG.Wait()
 }
 
 func TestLen(t *testing.T) {
@@ -158,4 +192,132 @@ func TestReinsert(t *testing.T) {
 	if a := q.Len(); a != 0 {
 		t.Errorf("Expected queue to be empty. Has %v items", a)
 	}
+}
+
+func TestQueueDrainageUsingShutDownWithDrain(t *testing.T) {
+
+	q := workqueue.New()
+
+	q.Add("foo")
+	q.Add("bar")
+
+	firstItem, _ := q.Get()
+	secondItem, _ := q.Get()
+
+	finishedWG := sync.WaitGroup{}
+	finishedWG.Add(1)
+	go func() {
+		defer finishedWG.Done()
+		q.ShutDownWithDrain()
+	}()
+
+	// This is done as to simulate a sequence of events where ShutDownWithDrain
+	// is called before we start marking all items as done - thus simulating a
+	// drain where we wait for all items to finish processing.
+	shuttingDown := false
+	for !shuttingDown {
+		_, shuttingDown = q.Get()
+	}
+
+	// Mark the first two items as done, as to finish up
+	q.Done(firstItem)
+	q.Done(secondItem)
+
+	finishedWG.Wait()
+}
+
+func TestNoQueueDrainageUsingShutDown(t *testing.T) {
+
+	q := workqueue.New()
+
+	q.Add("foo")
+	q.Add("bar")
+
+	q.Get()
+	q.Get()
+
+	finishedWG := sync.WaitGroup{}
+	finishedWG.Add(1)
+	go func() {
+		defer finishedWG.Done()
+		// Invoke ShutDown: suspending the execution immediately.
+		q.ShutDown()
+	}()
+
+	// We can now do this and not have the test timeout because we didn't call
+	// Done on the first two items before arriving here.
+	finishedWG.Wait()
+}
+
+func TestForceQueueShutdownUsingShutDown(t *testing.T) {
+
+	q := workqueue.New()
+
+	q.Add("foo")
+	q.Add("bar")
+
+	q.Get()
+	q.Get()
+
+	finishedWG := sync.WaitGroup{}
+	finishedWG.Add(1)
+	go func() {
+		defer finishedWG.Done()
+		q.ShutDownWithDrain()
+	}()
+
+	// This is done as to simulate a sequence of events where ShutDownWithDrain
+	// is called before ShutDown
+	shuttingDown := false
+	for !shuttingDown {
+		_, shuttingDown = q.Get()
+	}
+
+	// Use ShutDown to force the queue to shut down (simulating a caller
+	// which can invoke this function on a second SIGTERM/SIGINT)
+	q.ShutDown()
+
+	// We can now do this and not have the test timeout because we didn't call
+	// done on any of the items before arriving here.
+	finishedWG.Wait()
+}
+
+func TestQueueDrainageUsingShutDownWithDrainWithDirtyItem(t *testing.T) {
+	q := workqueue.New()
+
+	q.Add("foo")
+	gotten, _ := q.Get()
+	q.Add("foo")
+
+	finishedWG := sync.WaitGroup{}
+	finishedWG.Add(1)
+	go func() {
+		defer finishedWG.Done()
+		q.ShutDownWithDrain()
+	}()
+
+	// Ensure that ShutDownWithDrain has started and is blocked.
+	shuttingDown := false
+	for !shuttingDown {
+		_, shuttingDown = q.Get()
+	}
+
+	// Finish "working".
+	q.Done(gotten)
+
+	// `shuttingDown` becomes false because Done caused an item to go back into
+	// the queue.
+	again, shuttingDown := q.Get()
+	if shuttingDown {
+		t.Fatalf("should not have been done")
+	}
+	q.Done(again)
+
+	// Now we are really done.
+	_, shuttingDown = q.Get()
+	if !shuttingDown {
+		t.Fatalf("should have been done")
+	}
+
+	finishedWG.Wait()
 }

